src/services/fungibleTokens.js

```diff
@@ -0,0 +1,81 @@
+import BN from 'bn.js';
+import * as nearApiJs from 'near-api-js';
+
+import sendJson from '../tmp_fetch_send_json';
+import { parseTokenAmount } from '../utils/amounts';
+import { ACCOUNT_HELPER_URL } from '../utils/wallet';
+
+const {
+    transactions: {
+        functionCall
+    },
+    utils: {
+        format: { 
+            parseNearAmount
+        }
+    }
+} = nearApiJs;
+
+// account creation costs 0.00125 NEAR for storage, 0.00000000003 NEAR for gas
+// https://docs.near.org/docs/api/naj-cookbook#wrap-and-unwrap-near
+const FT_MINIMUM_STORAGE_BALANCE = parseNearAmount('0.00125');
+const FT_STORAGE_DEPOSIT_GAS = parseNearAmount('0.00000000003');
+
+// set this to the same value as we use for creating an account and the remainder is refunded
+const FT_TRANSFER_GAS = parseNearAmount('0.00000000003');
+
+// contract might require an attached depositof of at least 1 yoctoNear on transfer methods
+// "This 1 yoctoNEAR is not enforced by this standard, but is encouraged to do. While ability to receive attached deposit is enforced by this token."
+// from: https://github.com/near/NEPs/issues/141
+const FT_TRANSFER_DEPOSIT = '1';
+
+// Fungible Token Standard
+// https://github.com/near/NEPs/tree/master/specs/Standards/FungibleToken
+export class FungibleTokens {
+    constructor(account) {
+        this.account = account;
+    }
+
+    async isStorageBalanceAvailable(contractName, accountId) {
+        return new BN((await this.getStorageBalance(contractName, accountId)).total).gte(new BN(FT_MINIMUM_STORAGE_BALANCE));
+    }
+
+    async getStorageBalance(contractName, accountId) {
+        return await this.account.viewFunction(contractName, 'storage_balance_of', { account_id: accountId });
+    }
+
+    async transfer({ contractMetadata: { contractName, decimals }, amount, receiverId, memo }) {
```

Comment:
Given your other comment that `contractMetadata` does not actually include `contractName`, I've been thinking about this a bit more -- what do you think about naming the first argument here 'token' instead?  

That way we don't have a concept of 'wallet contract metadata'  object that has a property on it that isn't _actually_ a part of the 'NEP141 token contract metadata' spec.

e.g.:
```javascript
async transfer({
  token: {
    contractName,
    metadata: {
      decimals
    }
  },
  amount,
  receiverId,
  memo
}) {...}
```

This way everything in `token.metadata` will exactly match the NEP141 spec, giving us a re-usable type like e.g. `NEP141TokenMetadata` when we get to TS, and `token.contractName` is just a string that is the name of the contract name used to interact with the token.

I'm thinking about this from a few angles:
1. If we start adding properties to `metadata` payloads, we run the risk of there being naming conflicts between what we have added and what the metadata endpoint returns in the future.
2. Our metadata won't match the spec, so anyone referencing the spec won't understand why there are properties on our metadata that aren't in the spec.
3. It will be tricky to work around the fact that there are two different object shapes for things that are both "metadata" -- and which one you are working with will depend entirely on where in the codebase we're working and whether or not we've mutated the metadata yet to add `contractName`.  By naming it differently it'll be easier to understand whether we're working with the `{ contractName, metadata }` version or just `metadata` returned from the NEP141 contract endpoint.

What do you think?

Reply:
100% agree. Done.